src/store.rs

```diff
@@ -540,6 +574,152 @@ pub fn write(
+        Location::Volatile => store.vfs().rename(from_path, to_path),
+    }
+    .map_err(|_| Error::FilesystemWriteFailure)
+}
+
+pub fn move_file(
+    store: impl Store,
+    from_location: Location,
+    from_path: &Path,
+    to_location: Location,
+    to_path: &Path,
+) -> Result<(), Error> {
+    debug_now!(
+        "Moving {:?}({}) to {:?}({})",
+        from_location,
+        from_path,
+        to_location,
+        to_path
+    );
+    match to_location {
+        Location::Internal => create_directories(store.ifs(), to_path),
+        Location::External => create_directories(store.efs(), to_path),
+        Location::Volatile => create_directories(store.vfs(), to_path),
+    }
+    .map_err(|_err| {
+        error!("Failed to create directories chunks: {:?}", _err);
+        Error::FilesystemWriteFailure
+    })?;
+
+    let on_fail = |_err| {
+        error!("Failed to rename file: {:?}", _err);
+        Error::FilesystemWriteFailure
+    };
+    // Fast path for same-filesystem
+    match (from_location, to_location) {
+        (Location::Internal, Location::Internal) => {
+            return store.ifs().rename(from_path, to_path).map_err(on_fail)
+        }
+        (Location::External, Location::External) => {
+            return store.efs().rename(from_path, to_path).map_err(on_fail)
+        }
+        (Location::Volatile, Location::Volatile) => {
+            return store.vfs().rename(from_path, to_path).map_err(on_fail)
+        }
+        _ => {}
+    }
+
+    match from_location {
+        Location::Internal => {
+            move_opened_file(store, &**store.ifs(), from_path, to_location, to_path)
+        }
+        Location::External => {
+            move_opened_file(store, &**store.efs(), from_path, to_location, to_path)
+        }
+        Location::Volatile => {
+            move_opened_file(store, &**store.vfs(), from_path, to_location, to_path)
+        }
+    }
+}
+
+fn move_opened_file<S: LfsStorage>(
+    store: impl Store,
+    from_fs: &Filesystem<S>,
+    from_path: &Path,
+    to_location: Location,
+    to_path: &Path,
+) -> Result<(), Error> {
+    match to_location {
+        Location::Internal => move_to_opened_file(from_fs, from_path, &**store.ifs(), to_path),
+        Location::External => move_to_opened_file(from_fs, from_path, &**store.efs(), to_path),
+        Location::Volatile => move_to_opened_file(from_fs, from_path, &**store.vfs(), to_path),
+    }
+}
+
+fn move_to_opened_file<S1: LfsStorage, S2: LfsStorage>(
+    from_fs: &Filesystem<S1>,
+    from_path: &Path,
+    to_fs: &Filesystem<S2>,
+    to_path: &Path,
+) -> Result<(), Error> {
+    File::open_and_then(from_fs, from_path, |from_file| {
+        File::create_and_then(to_fs, to_path, |to_file| move_files(from_file, to_file))
+    })
+    .map_err(|_err| {
+        error!("Failed to flush chunks: {:?}", _err);
+        Error::FilesystemWriteFailure
+    })
+}
+
+fn move_files<S1: LfsStorage, S2: LfsStorage>(
```

Comment:
naming detail: should this be singular? so `move_file` ?

Reply:
These functions are hard to name :sweat_smile: . They're here mostly to avoid having 9 different branches for all combinations of `to` and `from` locations, because the types are distinct (so we need to use generics).

I renamed them to `move_file_step1`, `move_file_step2`, and `copy_file_data`, and added comments explaining  why. This documents intent better than trying to find a real name for the functions.